### PR TITLE
Users/ajya/WFC and powershell on target machine to ps3

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 1]
```diff
@@ -1,175 +1,178 @@
...
-    $resourcesPropertyBag = Get-ResourcesProperties -resources $resources
-}
-catch
-{
-    Write-Telemetry "Task_InternalError" $_.exception.Message
-
-    throw
-}
+    # Telemetry
+    Import-Module $PSScriptRoot\ps_modules\TelemetryHelper
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: keep all import at once place if possible

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 2]
```diff
@@ -1,175 +1,178 @@
...
+    if($runPowershellInParallel -eq "false" -or  ( $resources.Count -eq 1 ) )
+    {
+        foreach($resource in $resources)
+        {
+            $resourceProperties = $resourcesPropertyBag.Item($resource.Id)
+            $machine = $resourceProperties.fqdn
+            $displayName = $resourceProperties.displayName
+            Write-Output (Get-VstsLocString -Key "PS_TM_DeploymentStartedForMachine0" -ArgumentList $displayName)
+
+            $deploymentResponse = Invoke-Command -ScriptBlock $RunPowershellJob -ArgumentList $machine, $scriptPath, $resourceProperties.winrmPort, $scriptArguments, $initializationScriptPath, $resourceProperties.credential, $resourceProperties.protocolOption, $resourceProperties.skipCACheckOption, $enableDetailedLoggingString, $sessionVariables
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: ensure you are not doing any import from agent set-up in job.ps1. All the required imports should be form $psroot or task folder only

File: Tasks/PowerShellOnTargetMachines/task.json [hunk 3]
```diff
@@ -11,9 +11,9 @@
     ],
     "author": "Microsoft Corporation",
     "version": {
-        "Major": 1,
+        "Major": 2,
         "Minor": 0,
-        "Patch": 47
+        "Patch": 0
     },
     "minimumAgentVersion": "1.104.0",
```
Comment (chshrikh) [inline Tasks/PowerShellOnTargetMachines/task.json]: min agent version should be 2.* as v1 does not support PS3 handler. Better to keep this as latest released agent

File: Tasks/WindowsMachineFileCopy/task.json [hunk 4]
```diff
@@ -11,9 +11,9 @@
     ],
     "author": "Microsoft Corporation",
     "version": {
-        "Major": 1,
+        "Major": 2,
         "Minor": 0,
-        "Patch": 41
+        "Patch": 0
     },
     "minimumAgentVersion": "1.104.0",
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/task.json]: min agent version should be for V2

File: Tasks/WindowsMachineFileCopy/Utility.ps1 [hunk 5]
```diff
@@ -61,7 +61,7 @@ function Validate-Null(
     $value = $value.Trim()
     if(-not $value)
     {
-        ThrowError -errorMessage (Get-LocalizedString -Key "Parameter '{0}' cannot be null or empty." -ArgumentList $variableName)
+        ThrowError -errorMessage (Get-VstsLocString -Key "WFC_ParameterCannotBeNullorEmpty" -ArgumentList $variableName)
     }
 }
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/Utility.ps1]: IMO this file is not required at all as now we should not deal with any environment objects here

File: Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1 [hunk 6]
```diff
@@ -37,8 +37,6 @@ param (
         }
     }
     
-    import-module "Microsoft.TeamFoundation.DistributedTask.Task.Common"
-    
     function ThrowError
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1]: now windowsfilecopyjob.ps1 is not required. please remove it

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 7]
```diff
@@ -1,122 +1,108 @@
...
-Write-Verbose "additionalArguments = $additionalArguments"
-Write-Verbose "copyFilesInParallel = $copyFilesInParallel"
-Write-Verbose "cleanTargetBeforeCopy = $cleanTargetBeforeCopy"
+[CmdletBinding()]
+param()
+
+Trace-VstsEnteringInvocation $MyInvocation
+
+# Get inputs for the task
+$environmentName = Get-VstsInput -Name EnvironmentName
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: why we need environmentName ? we can work only on machineNames

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 8]
```diff
@@ -1,122 +1,108 @@
...
+$adminUserName = Get-VstsInput -Name AdminUserName
+$adminPassword = Get-VstsInput -Name AdminPassword
+$resourceFilteringMethod = Get-VstsInput -Name ResourceFilteringMethod
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: we dot not need to support filtering now . remove resourceFilteringMethod

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 9]
```diff
@@ -1,122 +1,108 @@
...
+    $envOperationStatus = 'Passed'
 
-    Write-Verbose "Starting Get-EnvironmentResources cmdlet call on environment name: $fetchedEnvironmentName"
-    $resources = Get-EnvironmentResources -Environment $environment
-    Write-Verbose "Completed Get-EnvironmentResources cmdlet call for environment name: $fetchedEnvironmentName"
+    Validate-SourcePath $sourcePath
+    Validate-DestinationPath $targetPath $environmentName
 
-    if ($resources.Count -eq 0)
+    if([string]::IsNullOrWhiteSpace($environmentName))
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: change environmentName to machine names

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 10]
```diff
@@ -1,122 +1,108 @@
...
-    {
-        foreach($resource in $resources)
+        $secureAdminPassword = ConvertTo-SecureString $adminPassword -AsPlainText -Force
+        $machineCredential = New-Object System.Net.NetworkCredential ($adminUserName, $secureAdminPassword)
+
+        if ($machines.Count -eq 0)
         {
-            $resourceProperties = $resourcesPropertyBag.Item($resource.Id)
-            $machine = $resourceProperties.fqdn        
+            throw (Get-VstsLocString -Key "WFC_NoMachineExistsUnderEnvironment0ForDeployment" -ArgumentList $environmentName)
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: move this check before working on creds

File: Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1 [hunk 11]
```diff
@@ -37,8 +37,6 @@ param (
         }
     }
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1]: ensure we do not import any file /module from agent. We should import it from task folder only

File: Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1 (hunk 6 above)
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1]: ThrowError function need to throw the string from loc file

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 12]
```diff
@@ -1,122 +1,108 @@
...
+        Write-Output (Get-VstsLocString -Key "Copy started for - '{0}'" -ArgumentList $targetPath)
+        Copy-OnLocalMachine -sourcePath $sourcePath -targetPath $targetPath -adminUserName $adminUserName -adminPassword $adminPassword `
+                            -cleanTargetBeforeCopy $cleanTargetBeforeCopy -additionalArguments $additionalArguments
+        Write-Verbose "Files copied to destination successfully."
     }
+    else
+    {
 
-    $resourcesPropertyBag = Get-ResourcesProperties -envName $fetchedEnvironmentName -resources $resources
+        $machines = $environmentName.split(',') | ForEach-Object { if ($_ -and $_.trim()) { $_.trim() } }
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: will be better if we log for machine names like - copy will be done for machines - m1,m2,m3

File: Tasks/WindowsMachineFileCopy/Utility.ps1 (hunk 5 above)
Comment (Ajay-MS) [reply Tasks/WindowsMachineFileCopy/Utility.ps1]: Removed unnecessary methods

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 13]
```diff
@@ -1,175 +1,177 @@
...
-Write-Verbose "runPowershellInParallel = $runPowershellInParallel"
-Write-Verbose "sessionVariables = $sessionVariables"
+[CmdletBinding()]
+param()
+
+Trace-VstsEnteringInvocation $MyInvocation
+
+# Get inputs for the task
+$environmentName = Get-VstsInput -Name EnvironmentName
+$adminUserName = Get-VstsInput -Name AdminUserName
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Should this be a required input?

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 14]
```diff
@@ -1,175 +1,177 @@
...
-
-    Write-Verbose "Starting Register-Environment cmdlet call for environment : $environmentName with filter $machineFilter"
-    $environment = Register-Environment -EnvironmentName $environmentName -EnvironmentSpecification $environmentName -UserName $adminUserName -Password $adminPassword -WinRmProtocol $protocol -TestCertificate ($testCertificate -eq "true") -Connection $connection -TaskContext $distributedTaskContext -ResourceFilter $machineFilter
-    Write-Verbose "Completed Register-Environment cmdlet call for environment : $environmentName"
-
-    Write-Verbose "Starting Get-EnvironmentResources cmdlet call on environment name: $environmentName"
-    $resources = Get-EnvironmentResources -Environment $environment
-
-    if ($resources.Count -eq 0)
+    # keep machineNames parameter name unchanged due to back compatibility
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: What is back compat in this new version of the task?

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 [hunk 15]
```diff
@@ -1,175 +1,177 @@
...
+    $machineFilter = $machineNames
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: What is machineFilter now? Do we need this?

File: Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1 (hunk 13 above)
Comment (kmkumaran) [reply Tasks/PowerShellOnTargetMachines/PowerShellOnTargetMachines.ps1]: Should you also use this rev of task to validate if the user input is in proper format?

File: Tasks/PowerShellOnTargetMachines/make.json [hunk 16]
```diff
@@ -4,5 +4,34 @@
...
+    "externals": {
+        "archivePackages": [
+            {
+                "url": "https://vstsagenttools.blob.core.windows.net/tools/DeploymentUtilities/M118.1/DeploymentUtilities.zip",
+                "dest": "./"
+            }
+        ],
+        "nugetv2": [
+            {
+                "name": "VstsTaskSdk",
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/make.json]: why do you need this?

File: Tasks/PowerShellOnTargetMachines/task.json [hunk 17]
```diff
@@ -11,11 +11,11 @@
...
     "author": "Microsoft Corporation",
     "version": {
-        "Major": 1,
+        "Major": 2,
         "Minor": 0,
-        "Patch": 47
+        "Patch": 0
     },
-    "minimumAgentVersion": "1.104.0",
+    "minimumAgentVersion": "2.120.0",
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/task.json]: Why is this requirement?

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 18]
```diff
@@ -1,122 +1,104 @@
...
-$machineFilter = $machineNames
-$sourcePath = $sourcePath.Trim('"')
-$targetPath = $targetPath.Trim('"')
+Trace-VstsEnteringInvocation $MyInvocation
 
-# Default + constants #
-$resourceFQDNKeyName = Get-ResourceFQDNTagKey
+# Get inputs for the task
+$machineNames = Get-VstsInput -Name MachineNames
+$adminUserName = Get-VstsInput -Name AdminUserName
```
Comment (kmkumaran) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: Should we make username/password mandatory now? And should we check if the format is proper for user name?

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 19]
```diff
@@ -1,122 +1,104 @@
...
-Validate-SourcePath $sourcePath
-Validate-DestinationPath $targetPath $environmentName
+. $PSScriptRoot/RoboCopyJob.ps1
+. $PSScriptRoot/Utility.ps1
 
-if([string]::IsNullOrWhiteSpace($environmentName))
+try 
 {
-    Write-Verbose "No environment found. Copying to destination."
+    # keep machineNames parameter name unchanged due to back compatibility
```
Comment (kmkumaran) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: Similar comment - not sure what the compatibility means now.

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 20]
```diff
@@ -1,122 +1,104 @@
...
-
-    Write-Verbose "Starting Get-EnvironmentResources cmdlet call on environment name: $fetchedEnvironmentName"
-    $resources = Get-EnvironmentResources -Environment $environment
-    Write-Verbose "Completed Get-EnvironmentResources cmdlet call for environment name: $fetchedEnvironmentName"
-
-    if ($resources.Count -eq 0)
+    if([string]::IsNullOrWhiteSpace($machineNames))
     {
-         throw (Get-LocalizedString -Key "No machine exists under environment: '{0}' for deployment" -ArgumentList $environmentName)
+        Write-Verbose "No environment found. Copying to destination."
```
Comment (kmkumaran) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: No *Machines* found....

Comment (kmkumaran) [reply Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: What is this scenario where we copy to local machine when machine name is empty. I thought we will make machine names a mandatory parameter.

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopyJob.ps1 [hunk 21]
```diff
@@ -1,38 +0,0 @@
-$CopyJob = {
```
Comment (kmkumaran) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopyJob.ps1]: For copying in parallel the main script still seem to use CopyJob script. Now that you are removing here, where is the script coming from?

File: Tasks/WindowsMachineFileCopy/task.json (hunk 4 above)
Comment (kmkumaran) [reply Tasks/WindowsMachineFileCopy/task.json]: Why V2?

File: Tasks/WindowsMachineFileCopy/task.json [hunk 22]
```diff
@@ -11,9 +11,9 @@
     ],
     "author": "Microsoft Corporation",
     "version": {
-        "Major": 1,
+        "Major": 2,
         "Minor": 0,
-        "Patch": 41
+        "Patch": 0
```
Comment (kmkumaran) [inline Tasks/WindowsMachineFileCopy/task.json]: You need to add "releaseNotes": attribute when you rev up major version.

File: Tasks/PowerShellOnTargetMachines/task.json [hunk 23]
```diff
@@ -11,11 +11,11 @@
     ],
     "author": "Microsoft Corporation",
     "version": {
-        "Major": 1,
+        "Major": 2,
```
Comment (kmkumaran) [inline Tasks/PowerShellOnTargetMachines/task.json]: You need to add "releaseNotes": attribute when you rev up major version.

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopyJob.ps1 (hunk 21 above)
Comment (Ajay-MS) [reply Tasks/WindowsMachineFileCopy/WindowsMachineFileCopyJob.ps1]: I didn't find any usage of this script . $copy job is being used from RoboCopyJob.ps1

File: Tasks/PowerShellOnTargetMachines/make.json (hunk 16 above)
Comment (Ajay-MS) [reply Tasks/PowerShellOnTargetMachines/make.json]: PS3 releated utility methods are in this module e.g Find-vstsfiles,Get-VstsLocString etc  so it's required for PS3.

File: Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1 [hunk 24]
```diff
@@ -1,106 +1,72 @@
...
 
-# keep machineNames parameter name unchanged due to back compatibility
-$machineFilter = $machineNames
-$sourcePath = $sourcePath.Trim('"')
-$targetPath = $targetPath.Trim('"')
+# Get inputs for the task
+$machineNames = Get-VstsInput -Name MachineNames -Require
+$adminUserName = Get-VstsInput -Name AdminUserName -Require
+$adminPassword = Get-VstsInput -Name AdminPassword -Require
+$sourcePath = Get-VstsInput -Name SourcePath
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/WindowsMachineFileCopy.ps1]: make sourcePath and targetPath both as required

File: Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1 [hunk 25]
```diff
@@ -21,23 +21,6 @@ param (
         $sourceDirectory = Split-Path $sourcePath
         $filesToCopy = Split-Path $sourcePath -Leaf
```
Comment (chshrikh) [inline Tasks/WindowsMachineFileCopy/RoboCopyJob.ps1]: add the logging for input parameters of job

**Pull request merged**